<launch>
  
  <!-- Load hand geometry parameters -->  
  <include file="$(find gpd)/launch/hand_geometry.launch">
    <arg name="node" value="detect_grasps" />
  </include>
  
  <!-- Load classifier parameters -->
  <include file="$(find gpd)/launch/lenet/ur5_classifier_15channels.launch">
    <arg name="node" value="detect_grasps" />
  </include>
  
	<node name="detect_grasps" pkg="gpd" type="detect_grasps" output="screen">
    
    <!-- If sequential importance sampling is used (default: false) -->
    <param name="use_importance_sampling" value="false" />
        
    <!-- What type of point cloud is used and what ROS topic it comes from -->
    <param name="cloud_type" value="1" /> <!-- 0: PointCloud2, 1: CloudIndexed, 2: CloudSamples -->
    <param name="cloud_topic" value="/cloud_indexed" />
    
    <!-- (optional) The ROS topic that the samples come from (default: an empty string) -->
    <param name="samples_topic" value="" />
    
    <!-- Plotting parameters -->
    <param name="plot_normals" value="false" />
    <param name="plot_samples" value="false" />    
    <param name="plot_candidates" value="false" />
    <param name="plot_filtered_grasps" value="false" />
    <param name="plot_valid_grasps" value="false" />
    <param name="plot_clusters" value="false" />
    <param name="plot_selected_grasps" value="false" />
    <param name="rviz_topic" value="grasps_rviz" />
    
    <!-- Preprocessing of point cloud -->
    <param name="voxelize" value="true"/>
    <param name="remove_outliers" value="false"/>
    <rosparam param="workspace"> [-1, 1, -1, 1, -1, 1] </rosparam>
    <rosparam param="camera_position"> [0, 0, 0] </rosparam>
            
    <!-- General parameters -->
    <param name="num_samples" value="100" />
    <param name="num_threads" value="4" />
    
    <!-- Parameters for local grasp candidate search -->
    <param name="nn_radius" value="0.01" />
    <param name="num_orientations" value="8" /> <!-- Number of orientations to consider -->
    
    <!-- Filtering of grasp candidates --> 
    <param name="filter_grasps" value="false" /> <!-- on workspace and robot hand aperture -->
    <rosparam param="workspace_grasps"> [0.55, 1.0, -0.41, 0.03, -0.29, 1.0] </rosparam>
    <param name="filter_half_antipodal" value="false"/> <!-- on half antipodal -->
    
    <!-- Filtering of grasps which are too low on (i.e. too close to) table -->
    <param name="filter_table_side_grasps" value="false" />
    <rosparam param="vertical_axis"> [0, 0, 1] </rosparam>
    <param name="angle_thresh" value="0.1" />
    <param name="table_height" value="0.0" />
    <param name="table_thresh" value="0.05" />
    
    <!-- Grasp image creation -->
    <param name="create_image_batches" value="false" /> <!-- creates grasp images in batches (less memory usage) -->
    <param name="remove_plane_before_image_calculation" value="false" /> <!-- removes table plane from point cloud to speed up shadow computations -->
    
    <!-- Clustering of grasps -->
    <param name="min_inliers" value="1" />
        
    <!-- Grasp selection -->
    <param name="min_score_diff" value="0" />
    <param name="min_aperture" value="0.029" />
    <param name="max_aperture" value="0.072" />
    <param name="num_selected" value="5" />
        
	</node>
  
</launch>
